packages/mcp/src/server.ts

```diff
@@ -0,0 +1,253 @@
+   * @returns Converted tools registry
+   */
+  private convertTools(tools: ToolsInput): Record<string, ConvertedTool> {
+    const convertedTools: Record<string, ConvertedTool> = {};
+    for (const toolName of Object.keys(tools)) {
+      let inputSchema: any;
+      let zodSchema: z.ZodTypeAny;
+      const toolInstance = tools[toolName];
+      if (!toolInstance) {
+        void logger.warning(`Tool instance for '${toolName}' is undefined. Skipping.`);
+        continue;
+      }
+      if (typeof toolInstance.execute !== 'function') {
+        void logger.warning(`Tool '${toolName}' does not have a valid execute function. Skipping.`);
+        continue;
+      }
+      // Vercel tools: .parameters is either Zod or JSON schema
+      if (isVercelTool(toolInstance)) {
+        if (isZodType(toolInstance.parameters)) {
+          zodSchema = toolInstance.parameters;
+          inputSchema = zodToJsonSchema(zodSchema);
+        } else if (typeof toolInstance.parameters === 'object') {
+          zodSchema = resolveSerializedZodOutput(jsonSchemaToZod(toolInstance.parameters));
+          inputSchema = toolInstance.parameters;
+        } else {
+          zodSchema = z.object({});
+          inputSchema = zodToJsonSchema(zodSchema);
+        }
+      } else {
+        // Mastra tools: .inputSchema is always Zod
+        zodSchema = toolInstance?.inputSchema ?? z.object({});
+        inputSchema = zodToJsonSchema(zodSchema);
+      }
+
+      // Wrap execute to support both signatures (typed, returns Promise<any>)
+      const execute: (args: any, execOptions?: any) => Promise<any> = async (args, execOptions) => {
+        if (isVercelTool(toolInstance)) {
+          return (await toolInstance.execute?.(args, execOptions)) ?? undefined;
+        }
+        return (await toolInstance.execute?.({ context: args }, execOptions)) ?? undefined;
+      };
+      convertedTools[toolName] = {
+        name: toolName,
+        description: toolInstance?.description,
+        inputSchema,
+        zodSchema,
+        execute,
+      };
+      void logger.info(`Registered tool: '${toolName}' [${toolInstance?.description || 'No description'}]`);
+    }
+    void logger.info(`Total tools registered: ${Object.keys(convertedTools).length}`);
+    return convertedTools;
+  }
+
+  /**
+   * Register the ListTools handler for listing all available tools.
+   */
+  private registerListToolsHandler() {
+    this.server.setRequestHandler(ListToolsRequestSchema, async () => {
+      await logger.debug('Handling ListTools request');
+      return {
+        tools: Object.values(this.convertedTools).map(tool => ({
+          name: tool.name,
+          description: tool.description,
+          inputSchema: tool.inputSchema,
+        })),
+      };
+    });
+  }
+
+  /**
+   * Register the CallTool handler for executing a tool by name.
+   */
+  private registerCallToolHandler() {
+    this.server.setRequestHandler(CallToolRequestSchema, async request => {
```

Comment:
Probably worth adding a bool on the class instance that we set/check to make sure we only add these handlers once per class instance